### PR TITLE
CLN: use idiomatic pandas_dtypes in pandas/dtypes/common.py

### LINKED ISSUE #24593: COMPAT: IntegerXXDtypes string repr cause numpy deprecation warnings
The 'numeric-style' type codes are deprecated in numpy, but they clash with our pandas integer dtypes strings (with the capital):

```
In [7]: np.dtype('Int64')                                     
/home/joris/miniconda3/envs/dev/bin/ipython:1: DeprecationWarning: Numeric-style type codes are deprecated and will result in an error in the future.
  #!/home/joris/miniconda3/envs/dev/bin/python
Out[7]: dtype('int64')
```

This causes some deprecation warnings in `pd.core.dtypes.common._get_dtype_type`, which surfaces those through eg `is_integer_dtype`:
```
In [1]: import warnings                                      

In [2]: warnings.simplefilter('always', DeprecationWarning)

In [3]: pd.api.types.pandas_dtype('Int64')                                                
Out[3]: <pandas.core.arrays.integer.Int64Dtype at 0x7fa8dc28c518>

In [4]: pd.core.dtypes.common._get_dtype_type('Int64')                                             
/home/joris/scipy/pandas/pandas/core/dtypes/common.py:1864: DeprecationWarning: Numeric-style type codes are deprecated and will result in an error in the future.
  return _get_dtype_type(np.dtype(arr_or_dtype))
Out[4]: numpy.int64

In [6]: pd.api.types.is_integer_dtype('Int64')                         
/home/joris/scipy/pandas/pandas/core/dtypes/common.py:1864: DeprecationWarning: Numeric-style type codes are deprecated and will result in an error in the future.
  return _get_dtype_type(np.dtype(arr_or_dtype))
Out[6]: True
```

### PR DESCRIPTION
**Pull request by jreback**
closes #24593 

Some benchmarks of ``pandas_dtype`` construction from a dtype object & strings, only thing slightly suprising is ``Period[D]``
```
[ 87.50%] ··· ============================================ =============
                                 dtype                                  
              -------------------------------------------- -------------
                             dtype('int64')                   463±20ns  
                             dtype('int32')                   450±20ns  
                            dtype('uint32')                   444±4ns   
                            dtype('uint64')                   484±20ns  
                            dtype('float32')                  494±30ns  
                            dtype('float64')                  507±30ns  
                             dtype('int16')                   471±30ns  
                             dtype('int8')                    506±30ns  
                            dtype('uint16')                   505±40ns  
                             dtype('uint8')                   485±20ns  
                              dtype('<M8')                    451±20ns  
                              dtype('<m8')                    480±20ns  
                               dtype('O')                    627±200ns  
                  pandas.core.arrays.integer.Int8Dtype        995±60ns  
                 pandas.core.arrays.integer.Int16Dtype        944±60ns  
                 pandas.core.arrays.integer.Int32Dtype        939±70ns  
                 pandas.core.arrays.integer.Int64Dtype        988±30ns  
                 pandas.core.arrays.integer.UInt8Dtype        924±50ns  
                 pandas.core.arrays.integer.UInt16Dtype       954±60ns  
                 pandas.core.arrays.integer.UInt32Dtype       966±70ns  
                 pandas.core.arrays.integer.UInt64Dtype     1.00±0.06μs 
               pandas.core.dtypes.dtypes.CategoricalDtype     978±30ns  
                pandas.core.dtypes.dtypes.IntervalDtype      929±300ns  
                          datetime64[ns, UTC]               1.52±0.03μs 
                               period[D]                      958±9ns   
                                 int64                       16.1±0.1μs 
                                 int32                       16.1±0.6μs 
                                 uint32                      16.0±0.1μs 
                                 uint64                      16.2±0.3μs 
                                float32                      15.8±0.4μs 
                                float64                      15.7±0.6μs 
                                 int16                       16.0±0.2μs 
                                  int8                      15.7±0.06μs 
                                 uint16                     15.9±0.08μs 
                                 uint8                       17.3±0.6μs 
                               datetime64                    16.2±0.3μs 
                              timedelta64                    16.0±0.1μs 
                                 object                      16.0±0.2μs 
                                  Int8                       6.24±0.1μs 
                                 Int16                       7.41±0.2μs 
                                 Int32                      8.15±0.04μs 
                                 Int64                       9.39±0.1μs 
                                 UInt8                       10.1±0.1μs 
                                 UInt16                     11.0±0.04μs 
                                 UInt32                      11.8±0.5μs 
                                 UInt64                     12.7±0.08μs 
                                category                    2.62±0.03μs 
                                interval                     6.72±0.2μs 
                          datetime64[ns, UTC]                2.93±0.1μs 
                               period[D]                     52.6±0.6μs 
              ============================================ =============
```

### CODE REVIEW

**Comment by jreback** [comment]
cc @TomAugspurger @jbrockmendel 

should be orthogonal to anything you are doing.

File: pandas/core/internals/construction.py [hunk 1]
```diff
@@ -559,11 +559,12 @@ def sanitize_array(data, index, dtype=None, copy=False,
 
             # possibility of nan -> garbage
             if is_float_dtype(data.dtype) and is_integer_dtype(dtype):
```
Comment (jreback) [inline pandas/core/internals/construction.py]: I had to change this becuase ``is_integer_dtype`` picks up nullable-integer types, I think it was just buggy before.

**Comment by jreback** [comment]
@jbrockmendel if any comments

File: asv_bench/benchmarks/dtypes.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from pandas.api.types import pandas_dtype
+
+import numpy as np
+from .pandas_vb_common import (
+    numeric_dtypes, datetime_dtypes, string_dtypes, extension_dtypes)
+
+
+_numpy_dtypes = list(map(np.dtype, (numeric_dtypes +
+                                    datetime_dtypes +
+                                    string_dtypes)))
```
Comment (jbrockmendel) [inline asv_bench/benchmarks/dtypes.py]: can you do list comprehension

File: pandas/core/dtypes/common.py [hunk 3]
```diff
@@ -1761,10 +1713,33 @@ def is_complex_dtype(arr_or_dtype):
...
+
+    """
+
     if arr_or_dtype is None:
         return False
-    tipo = _get_dtype_type(arr_or_dtype)
-    return issubclass(tipo, np.complexfloating)
+    try:
+        dtype = _get_dtype(arr_or_dtype)
+    except (TypeError, ValueError, SyntaxError, UnicodeEncodeError):
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/dtypes/common.py]: yeah we had this somewhere in a specific test, let me seeif i can take it out

File: pandas/core/internals/construction.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/construction.py]: does this need a IntNA test then?

**Comment by jbrockmendel** [comment]
A few comments, one possible request for a test.  Otherwise LGTM.

**Comment by jreback** [comment]
updated

**Comment by jreback** [comment]
i have a couple of more changes here.

**Comment by jbrockmendel** [comment]
ill take a look

File: pandas/core/dtypes/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/dtypes/common.py]: so its still needed here?

File: pandas/core/dtypes/common.py [hunk 4]
```diff
@@ -1787,92 +1762,67 @@ def _get_dtype(arr_or_dtype):
...
 
-    # TODO(extension)
-    # replace with pandas_dtype
-
     if arr_or_dtype is None:
         raise TypeError("Cannot deduce dtype from null object")
-    if isinstance(arr_or_dtype, np.dtype):
+
+    # fastpath
+    elif isinstance(arr_or_dtype, np.dtype):
```
Comment (jbrockmendel) [inline pandas/core/dtypes/common.py]: you caught ExtensionDType etc in fastpath below.  is there a reason not to do it here?

File: pandas/core/indexes/numeric.py [hunk 5]
```diff
@@ -328,7 +328,9 @@ def astype(self, dtype, copy=True):
             msg = ('Cannot convert Float64Index to dtype {dtype}; integer '
                    'values are required for conversion').format(dtype=dtype)
             raise TypeError(msg)
-        elif is_integer_dtype(dtype) and self.hasnans:
+        elif (is_integer_dtype(dtype) and
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/numeric.py]: Did you add a test for this? 
(it is currently failing (meaning: raising an error) on master)

**Review by jorisvandenbossche — COMMENTED** [review]
Does this change the behaviour of `is_integer_dtype` (getting that from one of your comments). Currently that returns False for Integer64Dtype. Not sure we should change that (code could rely on the fact there would be no NaNs for integers).

File: pandas/core/indexes/numeric.py (hunk 5 above)
Comment (jreback) [reply pandas/core/indexes/numeric.py]: this is well tested in many locations

File: pandas/core/internals/construction.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/construction.py]: I think this is my only unresolved comment.

**Comment by jreback** [comment]
> Does this change the behaviour of is_integer_dtype (getting that from one of your comments). Currently that returns False for Integer64Dtype. Not sure we should change that (code could rely on the fact there would be no NaNs for integers).

yes this was broken before

**Comment by jreback** [comment]
@jbrockmendel I think you had another comment but can't find it.

File: pandas/core/dtypes/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/dtypes/common.py]: this is now removed

File: pandas/core/indexes/numeric.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/indexes/numeric.py]: Ah, yes, forgot about the object dtype of the result

**Comment by jorisvandenbossche** [comment]
> > Does this change the behaviour of is_integer_dtype (getting that from one of your comments). Currently that returns False for Integer64Dtype. Not sure we should change that (code could rely on the fact there would be no NaNs for integers).
>
> yes this was broken before

It could also have been a choice :)

In any case, can you then update the docstring of `is_integer_dtype` to reflect this change? (is `is_int64_dtype` also affected?)

**Comment by jreback** [comment]
ok this is ready. @jorisvandenbossche @jbrockmendel

**Comment by TomAugspurger** [comment]
So I think this closes https://github.com/pandas-dev/pandas/issues/24593 then?

File: pandas/core/dtypes/common.py [hunk 6]
```diff
@@ -116,6 +114,20 @@ def ensure_int64_or_float64(arr, copy=False):
         return arr.astype('float64', copy=copy)
 
 
+def _issubclass(*klasses):
```
Comment (TomAugspurger) [inline pandas/core/dtypes/common.py]: nitpick: we could call this `classes`, since it's not a keyword.

File: pandas/core/dtypes/common.py [hunk 7]
```diff
@@ -1787,95 +1800,70 @@ def _get_dtype(arr_or_dtype):
...
-    Get the type (NOT dtype) instance associated with
-    an array or dtype object.
+    Return a boolean if the condition is satisfied for the arr_or_dtype.
 
     Parameters
     ----------
     arr_or_dtype : array-like
-        The array-like or dtype object whose type we want to extract.
+        The array-like or dtype object whose dtype we want to extract.
+    condition : callable[Union[type(np.dtype), ExtensionDtypeType]]
```
Comment (TomAugspurger) [inline pandas/core/dtypes/common.py]: I think `type(np.dtype)` should just be `np.dtype`

```python
In [8]: type(np.dtype('int'))
Out[8]: numpy.dtype

```

File: pandas/core/dtypes/common.py [hunk 8]
```diff
@@ -832,6 +823,10 @@ def is_any_int_dtype(arr_or_dtype):
 
     This function is internal and should not be exposed in the public API.
 
+    null-able Integer support
```
Comment (TomAugspurger) [inline pandas/core/dtypes/common.py]: null-able -> Nullable.

Also, I think that should go inside the `versionchanged` directive, not above it.

**Comment by jreback** [comment]
yes this closes #24593 as well. will address comments later today.

**Comment by jorisvandenbossche** [comment]
> yes this closes #24593 as wel

There are still a bunch of those deprecation warnings in the last travis build in this PR (although maybe coming from another place, that I didn't check. But strictly speaking not yet closing ##24593 completely)

**Comment by jreback** [comment]
@jorisvandenbossche where? https://github.com/pandas-dev/pandas/pull/24541#issuecomment-451303579

**Comment by jorisvandenbossche** [comment]
Eg https://travis-ci.org/pandas-dev/pandas/jobs/474964674 (first entry of the last green travis build at the time I commented before) still has some "eprecationWarning: Numeric-style type codes are deprecated and will result in an error in the future." warnings in the log

File: pandas/core/dtypes/common.py [hunk 9]
```diff
@@ -823,9 +823,9 @@ def is_any_int_dtype(arr_or_dtype):
 
     This function is internal and should not be exposed in the public API.
 
-    null-able Integer support
+    .. versionchanged:: 0.24.0
 
-    .. versionadded:: 0.24.0
+       Nullable Integer support
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/common.py]: Can you put this in a full sentence? 

Something like "The nullable Integer dtypes (e.g. pandas.Int64Dtype) are also considered as integer by this function."

**Comment by jreback** [comment]
I updated the whatsnew about #21681 as well.

**Comment by jreback** [comment]
ok warnings fixed up. Basically we were missing paths in core/internals/concat.py and core/dtypes/conat.py for EA operations.

File: pandas/compat/__init__.py [hunk 10]
```diff
@@ -139,7 +139,7 @@ def lfilter(*args, **kwargs):
     Hashable = collections.abc.Hashable
     Iterable = collections.abc.Iterable
     Mapping = collections.abc.Mapping
-    MutableMapping = collections.abc.MutableMapping
+    MutableMapping = collections.MutableMapping
```
Comment (jorisvandenbossche) [inline pandas/compat/__init__.py]: I think the one with `abc` included was actually correct? (CI is failing now due the warning this raises)

File: doc/source/whatsnew/v0.24.0.rst [hunk 11]
```diff
@@ -432,7 +432,7 @@ Backwards incompatible API changes
 - The column order of the resultant :class:`DataFrame` from :meth:`MultiIndex.to_frame` is now guaranteed to match the :attr:`MultiIndex.names` order. (:issue:`22420`)
 - Incorrectly passing a :class:`DatetimeIndex` to :meth:`MultiIndex.from_tuples`, rather than a sequence of tuples, now raises a ``TypeError`` rather than a ``ValueError`` (:issue:`24024`)
 - :func:`pd.offsets.generate_range` argument ``time_rule`` has been removed; use ``offset`` instead (:issue:`24157`)
-- In 0.23.x, pandas would raise a ``ValueError`` on a merge of a numeric column (e.g. ``int`` dtyped column) and an ``object`` dtyped column (:issue:`9780`). We have re-enabled the ability to merge ``object`` and other dtypes (:issue:`21681`)
+- In 0.23.x, pandas would raise a ``ValueError`` on a merge of a numeric column (e.g. ``int`` dtyped column) and an ``object`` dtyped column, though only if it is a numeric column (:issue:`9780`). We have re-enabled the ability to merge ``object`` and other dtypes; pandas will sill raise on a merge between a numeric and an ``object`` dtyped coumn that is composed only of strings (:issue:`21681`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.24.0.rst]:
```suggestion
- In 0.23.x, pandas would raise a ``ValueError`` on a merge of a numeric column (e.g. ``int`` dtyped column) and an ``object`` dtyped column (:issue:`9780`). We have re-enabled the ability to merge ``object`` and other dtypes; pandas will still raise on a merge between a numeric and an ``object`` dtyped column that is composed only of strings (:issue:`21681`)
```

**Pull request merged**